pkg/agent/agentserver/backend_manager.go

```diff
@@ -0,0 +1,130 @@
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package agentserver
+
+import (
+	"fmt"
+	"math/rand"
+	"sync"
+	"time"
+
+	"k8s.io/klog"
+	"sigs.k8s.io/apiserver-network-proxy/proto/agent"
+)
+
+// BackendManager is an interface to manage backend connections, i.e.,
+// connection to the proxy agents.
+type BackendManager interface {
+	// Backend returns a single backend.
+	Backend() (agent.AgentService_ConnectServer, error)
+	// AllBackends returns all backends, categorized based on agentID.
+	AllBackends() map[string][]agent.AgentService_ConnectServer
+	// AddBackend adds a backend.
+	AddBackend(agentID string, conn agent.AgentService_ConnectServer)
+	// RemoveBackend removes a backend.
+	RemoveBackend(agentID string, conn agent.AgentService_ConnectServer)
+}
+
+var _ BackendManager = &DefaultBackendManager{}
+
+// DefaultBackendManager is the default backend manager.
+type DefaultBackendManager struct {
+	mu sync.RWMutex //protects the following
+	// A map between agentID and its grpc connections.
+	// For a given agent, ProxyServer prefers backends[agentID][0] to send
+	// traffic, because backends[agentID][1:] are more likely to be closed
+	// by the agent to deduplicate connections to the same server.
+	backends map[string][]agent.AgentService_ConnectServer
+	agentIDs []string
+	random   *rand.Rand
+}
+
+// NewDefaultBackendManager returns a DefaultBackendManager.
+func NewDefaultBackendManager() *DefaultBackendManager {
+	return &DefaultBackendManager{
+		backends: make(map[string][]agent.AgentService_ConnectServer),
+		random:   rand.New(rand.NewSource(time.Now().UnixNano())),
+	}
+}
+
+// AddBackend adds a backend.
+func (s *DefaultBackendManager) AddBackend(agentID string, conn agent.AgentService_ConnectServer) {
+	klog.Infof("register Backend %v for agentID %s", conn, agentID)
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	_, ok := s.backends[agentID]
+	if ok {
+		s.backends[agentID] = append(s.backends[agentID], conn)
+		return
+	}
+	s.backends[agentID] = []agent.AgentService_ConnectServer{conn}
+	s.agentIDs = append(s.agentIDs, agentID)
+}
+
+// RemoveBackend removes a backend.
+func (s *DefaultBackendManager) RemoveBackend(agentID string, conn agent.AgentService_ConnectServer) {
+	klog.Infof("remove Backend %v for agentID %s", conn, agentID)
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	backends, ok := s.backends[agentID]
+	if !ok {
+		klog.Warningf("can't find agentID %s in the backends", agentID)
+		return
+	}
+	var found bool
+	for i, c := range backends {
```

Comment:
Backend() assumes it will return element 0 of the slice. That means that if len(backends) > 0 and we remove i == 0 we have a problem. We should probably add a warning to that effect in the code.

Reply:
Added.